File destination flush signal

shikharvashistha opened the pull request:
<!--
Thank you for contributing to syslog-ng. Please make sure you:
- Read our Contribution guideline: https://github.com/syslog-ng/syslog-ng/blob/master/CONTRIBUTING.md
- Checked that tests pass (including stylechecks: `make style-check`)
- Wrote a news file, if applicable: https://github.com/syslog-ng/syslog-ng/tree/master/news
-->
This pull introduces file rotation mechanism to syslog-ng

<!--
PR description
In the description, please explain the problem your pull request intends to solve, and a give general overview of the implementation.
For more information, see: https://github.com/syslog-ng/syslog-ng/blob/master/CONTRIBUTING.md#pr-description
-->

MrAnno commented on modules/file-rotation/file-rotation-parser.c:
```diff
@@ -33,7 +33,7 @@ static CfgLexerKeyword file_rotation_keywords[] =
 {
   { "file_rotation", KW_FILE_ROTATION },
   { "size", KW_SIZE },
-  { NULL }
```
`NULL` marks the end of the list, it should not be removed.

MrAnno commented on modules/file-rotation/file-rotation.c:
```diff
@@ -27,6 +27,7 @@
 struct _FileRotationPlugin
 {
   LogDriverPlugin super;
+  char *interval;
```
Please use GLib types, wherever possible: `gchar *`

MrAnno commented on modules/file-rotation/file-rotation.c:
```diff
@@ -41,11 +42,21 @@ file_rotation_attach_to_driver(LogDriverPlugin *s, LogDriver *driver)
 file_rotation_free(LogDriverPlugin *s)
 {
-  FileRotationPlugin *self = (FileRotationPlugin *) s;
+  //FileRotationPlugin *self = (FileRotationPlugin *) s;
```
This does not belong in this commit.

Please make sure your commits are small atomic changes, where the commit message explains the whole change.

MrAnno commented on modules/file-rotation/file-rotation.h:
```diff
@@ -27,9 +27,11 @@
 FileRotationPlugin *file_rotation_new(void);
 
 void file_rotation_set_size(FileRotationPlugin *self, gsize size);
+void log_rotation_set_interval(LogDriver *self, char *interval);
```
We can use `FileRotationPlugin` as self. This plugin is not `LogDriver`-based.

MrAnno commented on modules/file-rotation/tests/test_file_rotation.c:
```diff
@@ -30,6 +30,7 @@ Test(file_rotation, test_example)
 {
   FileRotationPlugin *fr = file_rotation_new();
   file_rotation_set_size(fr, 100);
+  log_rotation_set_interval(fr, "1h");
```
I think we could use `logrotate`'s naming conventions here, for example:
daily, weekly, monthly, or yearly.

MrAnno commented on modules/affile/logproto-file-writer.c:
```diff
@@ -22,12 +22,7 @@
-#include "signal-slot-connector/signal-slot-connector.h"
-#include "modules/http/http-worker.h"
-#include "modules/http/http-signals.h"
-#include "modules/http/http.h"
```
We allow (encourage) faking/squashing/rebasing commit history while working on a feature branch.
The goal here is to create a nice linear patchset for your pull request, that is easy to review.

For the same reason, please use `git rebase` instead of merge-based pulls when updating your working branch.

MrAnno commented on modules/affile/logproto-file-writer.c:
```diff
@@ -153,16 +148,16 @@ log_proto_file_writer_flush(LogProtoClient *s)
+  FileRotationDriver *owner = (FileRotationDriver *) self->owner;
 
-  EMIT(owner->super.super.super.super.signal_slot_connector, signal_http_header_request, &signal_data);
+  EMIT(owner->super.super.super.super.signal_slot_connector, signal_file_rotation_request, &signal_data);
```
This probably does not work due to the above reasoning.

shikharvashistha replied:
So it should be http header request?

MrAnno replied:
No, it should be a `signal_file_rotation_request`, which needs to be declared in `file-signals.h`.

The problem here is that we need a `signal_slot_connector` instance, but not a completely new instance, we need `AFFileDestDriver`'s signal slot connector.
In order not to reference `AFFileDestDriver` directly from `LogProtoFileWriter` (unnecessary coupling), we could pass `AFFileDestDriver`'s signal slot connector to `log_proto_file_writer_new()` as a parameter.

shikharvashistha replied:
Fixed :+1:

MrAnno commented on modules/affile/logproto-file-writer.c:
```diff
@@ -36,7 +31,7 @@
 {
   LogProtoClient super;
-  HTTPDestinationWorker *worker;
+  LogThreadedDestWorker *owner;
```
We could store a `SignalSlotConnector *` pointer here, and it could be initialized by the constructor of LogProtoFileWriter: `log_proto_file_writer_new()`.

We need to pass the `AFFileDestDriver`'s `signal_slot_connector` instance to the constructor of LogProtoFileWriter.

shikharvashistha replied:
I think I'm doing this [here](https://imgur.com/a/9WXptpE)

shikharvashistha replied:
> I think I'm doing this [here](https://imgur.com/a/9WXptpE)

[commit-line](https://github.com/MrAnno/syslog-ng/blob/5f73c4f8922ce92afaa01817204f9fd35d8c0192/modules/affile/logproto-file-writer.c#L36)

MrAnno replied:
See my comment above. We need a specific instance from `AFFileDestDriver`, not a completely new one.

Also, please rename the field to `signal_slot_connector` in LogProtoFileWriter, as the connector is not the "owner" of the file writer.

shikharvashistha replied:
Updated :+1:

MrAnno commented on modules/affile/logproto-file-writer.h:
```diff
@@ -25,6 +25,15 @@
+typedef struct
+{
+  LogThreadedDestDriver super;
+} FileRotationDriver;
```
We don't need such a struct.

MrAnno commented on modules/affile/logproto-file-writer.h:
```diff
@@ -25,6 +25,15 @@
+#include "logthrdest/logthrdestdrv.h"
+#include "modules/file-rotation/file-rotation.h"
+
+typedef struct _FileRotationRequestData FileRotationRequestData;
```
Let's call this `FileFlushSignalData` for now.

The signal needs to be declared on this side, not in the file rotation plugin.
For that reason, let's call it `signal_file_flush`.

(We try to avoid mentioning file rotation here, we just provide generic signals, so any plugin can attach to them and do their job.)

All of these declarations should go into a separate header file with an LGPL license header.
We could call it `file-signals.h`.
An example can be found here: https://github.com/syslog-ng/syslog-ng/blob/master/modules/http/http-signals.h

shikharvashistha replied:
Created :+1:

MrAnno commented:
Our commit message template is the following:

```
modulename: short description of the commit

optional longer description

Signed-off-by: Name <email>
```

Marking bug/fix/chore is a good practice, but we start with the module's name instead.
In our case, `affile` and `file-rotation`.

shikharvashistha commented:
Hi, @MrAnno, I'm unable to see these conflicts locally don't know why these are there.

MrAnno commented:
I can't review this. Please fix the git history.
I'm happy to help with git rebase or anything, we can talk about it on Gitter.

MrAnno commented on modules/file-rotation/file-rotation-parser.c:
```diff
@@ -33,6 +33,7 @@ static CfgLexerKeyword file_rotation_keywords[] =
 {
   { "file_rotation", KW_FILE_ROTATION },
   { "size", KW_SIZE },
+  { "interval", KW_INTERVAL },
```
Adding the interval option does not belong to the flush-signal commit.

shikharvashistha replied:
I think these changes have been lost due to force push, what do you suggest should I separate out the commits ?

MrAnno replied:
Yes, please create a separate commit.

As I mentioned:
We encourage faking commit history while working on a feature branch.
The goal here is to create a nice linear patchset for your pull request, that is easy to review.
The commit message should describe the change it is containing, which needs to be small/atomic, concentrating on one single thing.

shikharvashistha replied:
Updated :+1:

MrAnno reviewed (commented):
I've also clicked "unresolve" on 3 of my previous notes. Those notes try to explain why this does not work.

Also, I can't compile the current version of this branch, `file-signals.h` is missing from this patchset, it is probably a rebase issue.

shikharvashistha commented:
> I've also clicked "unresolve" on 3 of my previous notes. Those notes try to explain why this does not work.
> 
> Also, I can't compile the current version of this branch, `file-signals.h` is missing from this patchset, it is probably a rebase issue.

Yeah because it didn't got commited used git add -u

MrAnno commented on modules/file-rotation/file-rotation.c:
```diff
@@ -22,29 +22,47 @@
 
 static gboolean
-file_rotation_attach_to_driver(LogDriverPlugin *s, LogDriver *driver)
+file_rotation_attach_to_driver(LogDriverPlugin *s, FileRotationPlugin *driver)
```
No need to change the signature of this function.

You can get a signal-slot-connector instance this way:

```
SignalSlotConnector *ssc = driver->super.signal_slot_connector;
```

(That instance will be the plugin's signal-slot-connector we attach to. In this example this is the `AFFileDestDriver`'s signal-slot-connector.)

shikharvashistha replied:
Changed :+1:

shikharvashistha committed:
affile: Integrated signal to the file destination

This commit adds signal to the file destination involving passing AFFileDestDriver's instance signal slot connector to the LogProtoFileWriter

Signed-off-by: Shikhar Vashistha <51105234+shikharvashistha@users.noreply.github.com>

shikharvashistha committed:
file-rotation: Added new interval option to file rotation plugin

This commit introduces new option to set interval for log rotation

Signed-off-by: Shikhar Vashistha <51105234+shikharvashistha@users.noreply.github.com>

shikharvashistha committed:
file-rotation: Added set_interval function & connected to slot

This commit introduces the controller function for updating/set log rotation interval & Adds support code for connecting AFFileDestDriver plugin's signal_slot_connector to our plugin

Signed-off-by: Shikhar Vashistha <51105234+shikharvashistha@users.noreply.github.com>

shikharvashistha committed:
affile: Seperated FileFlushData & SIGNAL to a new file  & Updated plugin test for set_interval

This commit updates the location of plugin's request payload and the SIGNAL which is required to be emitted

Signed-off-by: Shikhar Vashistha <51105234+shikharvashistha@users.noreply.github.com>

MrAnno commented on modules/affile/file-opener.h:
```diff
@@ -64,7 +65,8 @@ struct _FileOpener
                                          LogProtoFileReaderOptions *proto_options);
-  LogProtoClient *(*construct_dst_proto)(FileOpener *self, LogTransport *transport, LogProtoClientOptions *proto_options);
+  LogProtoClient *(*construct_dst_proto)(FileOpener *self, LogTransport *transport, LogProtoClientOptions *proto_options,
+                                         SignalSlotConnector *connector);
```
👍🏻 

The `construct_dst_proto` function pointer is for imitating C++ virtual methods in C.
Since you changed the signature of the virtual method, you need to adjust the parameter list in subclasses too.

The compiler warnings show you the missing pieces, you will eventually reach `log_proto_file_writer_new()`.

`modules/affile/named-pipe.c` - we don't want to do anything with the connector here, just fix the signature
`modules/affile/regular-files.c` - this is where it gets interesting

shikharvashistha replied:
Fixed :+1:

MrAnno commented on modules/affile/file-opener.h:
```diff
@@ -28,6 +28,7 @@
 #include "compat/lfs.h"
 
 #include "file-perms.h"
+#include "lib/signal-slot-connector/signal-slot-connector.h"
```
Let's just remove the `lib/` prefix, it should work too.

shikharvashistha replied:
Resolved

MrAnno commented on modules/affile/file-signals.h:
```diff
@@ -0,0 +1,30 @@
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 as published
+ * by the Free Software Foundation, or (at your option) any later version.
```
We should change the license of this file to LGPL in my opinion.

For example:
https://github.com/syslog-ng/syslog-ng/blob/master/lib/cache.h#L4-L20

Please add this file to the copyright checker script too:
https://github.com/syslog-ng/syslog-ng/blob/master/tests/copyright/policy#L119

You can test it by running `make check-copyright`.

shikharvashistha replied:
Updated :+1:

MrAnno commented on modules/affile/logproto-file-writer.c:
```diff
@@ -23,7 +23,8 @@
 #include "logproto-file-writer.h"
 #include "messages.h"
-
+#include "affile-dest.h"
```
That is not needed.

MrAnno commented on modules/affile/file-signals.h:
```diff
@@ -0,0 +1,30 @@
+
+#ifndef FILE_SIGNALS_H_INCLUDED
+#define FILE_SIGNALS_H_INCLUDED
+
```
#include "signal-slot-connector/signal-slot-connector.h"

shikharvashistha replied:
:+1:

MrAnno commented on modules/affile/logproto-file-writer.c:
```diff
@@ -32,6 +33,8 @@
 {
   LogProtoClient super;
+  SignalSlotConnector *signal_slot_connector;
+  AFFileDestDriver *signal_slot_connector_driver;
```
We should avoid storing or referencing `AFFileDestDriver` in this class, if possible.

shikharvashistha replied:
:+1:

MrAnno commented on modules/affile/logproto-file-writer.c:
```diff
@@ -134,6 +137,26 @@ log_proto_file_writer_flush(LogProtoClient *s)
+  //Register the signal to AFFileDestDriver's signal-slot-connector (the signal is emitted on the AFFileDestDriver's signal-slot-connector)
+
+  EMIT(self->signal_slot_connector, signal_file_flush, &signal_data);
+
```
Please remove all the comments here. We can get rid of the test data as well for now.
We can emit the signal with an empty `FileFlushSignalData`. The next task (and PR) will be about filling this data properly.

shikharvashistha replied:
:+1:

MrAnno commented on modules/affile/logproto-file-writer.c:
```diff
@@ -233,7 +255,11 @@ log_proto_file_writer_new(LogTransport *transport, const LogProtoClientOptions *
       struct iovec)*flush_lines);
 
+
+
```
Those whitespaces seem unnecessary.

shikharvashistha replied:
:+1:

MrAnno commented on modules/affile/logproto-file-writer.h:
```diff
@@ -25,8 +25,9 @@
 #define LOG_PROTO_FILE_WRITER_H_INCLUDED
 
 #include "logproto/logproto-client.h"
+#include "modules/file-rotation/file-rotation.h"
```
We should not include file-rotation in the affile module.

`#include "signal-slot-connector/signal-slot-connector.h"` should be enough here.

shikharvashistha replied:
:+1:

MrAnno commented on modules/file-rotation/file-rotation.c:
```diff
@@ -22,29 +22,46 @@
  */
 
 #include "file-rotation.h"
+#include "modules/http/http-signals.h"
```
We don't need HTTP signals here.

shikharvashistha replied:
:+1:

MrAnno commented on modules/file-rotation/file-rotation.c:
```diff
@@ -22,29 +22,46 @@
 
-struct _FileRotationPlugin
+static void
+_slot_append_file_rotation_request(FileRotationPlugin *self, FileFlushSignalData *data)
```
I'd call this function `_slot_file_rotation` or anything similar.

(We can't talk about "appending requests" here, those were just examples of the HTTP signaling).

shikharvashistha replied:
Updated :+1:

MrAnno commented on modules/file-rotation/file-rotation.c:
```diff
@@ -22,29 +22,46 @@
-  gsize size;
-};
+
+  g_string_append_printf(self->filename, "%s", data->filename->str);
```
I'd just add a `msg_trace()` here so you can test the slot by starting syslog-ng with `-Fdevt`.
There is no need to append the filename, we can just go with the msg_trace and nothing more for this PR.

shikharvashistha replied:
ok sure

MrAnno commented on modules/file-rotation/file-rotation.c:
```diff
@@ -22,29 +22,46 @@
-  msg_debug("File rotation plugin has been attached", evt_tag_str("to", driver->id));
+  SignalSlotConnector *ssc = driver->super.signal_slot_connector;
+
+  msg_debug("File rotation plugin has been attached", evt_tag_str("to", s->name));
```
The original `driver->id` was correct in my opinion, it showed the driver we attached our file-rotation plugin to.

shikharvashistha replied:
updated

MrAnno commented on modules/file-rotation/file-rotation.h:
```diff
@@ -25,11 +25,56 @@
+#include "signal-slot-connector/signal-slot-connector.h"
 
 typedef struct _FileRotationPlugin FileRotationPlugin;
+typedef struct _LogDriver LogDriver;
```
Please remove this.

shikharvashistha replied:
Removed

MrAnno commented on modules/file-rotation/file-rotation.h:
```diff
@@ -25,11 +25,56 @@
+  FILE_ROTATION_FAILURE,
+} FileRotationResult;
+
+struct _FileFlushSignalData
```
This should go into file-signals.h

shikharvashistha replied:
:+1:

MrAnno commented on modules/file-rotation/file-rotation.h:
```diff
@@ -25,11 +25,56 @@
+{
+  GString *filename;
+  gchar *interval;
+  gsize size;
```
filename can be useful, but we don't need interval and size here.

shikharvashistha replied:
Removed

MrAnno commented on modules/file-rotation/file-rotation.h:
```diff
@@ -25,11 +25,56 @@
+
+
+
+#define signal_file_rotation_response SIGNAL(file_rotation, rotation_response, FileRotationResponseData *)
```
This is not needed.

shikharvashistha replied:
:+1:

MrAnno commented on modules/file-rotation/file-rotation.h:
```diff
@@ -25,11 +25,56 @@
+{
+  FILE_ROTATION_INTERVAL,
+  FILE_ROTATION_SIZE,
+} FileRotationOption;
```
This is unused, let's remove it.

shikharvashistha replied:
Removed :+1:

MrAnno commented on modules/affile/logproto-file-writer.c:
```diff
@@ -134,6 +135,14 @@ log_proto_file_writer_flush(LogProtoClient *s)
+  GString *request_body = g_string_new("");
+  g_string_append_printf(request_body, "%d", self->buf_size);
+  GString *file_name = g_string_new("");
+  g_string_append_printf(file_name, "%s", self->super.transport->name);
```
`request_body` and `file_name` are not used currently. Let's remove them for now.

shikharvashistha replied:
Updated 👍

MrAnno commented on modules/file-rotation/file-rotation.c:
```diff
@@ -22,29 +22,44 @@
-  LogDriverPlugin super;
-  gsize size;
-};
+  msg_trace("File rotation signal received", evt_tag_str("filename", self->filename->str));
```
Does not this crash?
Currently, self->filename is probably a NULL pointer.

shikharvashistha replied:
Updated 👍

MrAnno reviewed (commented):
The proto constructor signature change hasn't been followed in our unit tests, please fix them there as well.

https://github.com/MrAnno/syslog-ng/runs/8036247829?check_suite_focus=true

MrAnno commented on modules/file-rotation/file-rotation-grammar.ym:
```diff
@@ -80,6 +81,7 @@ file_rotation_options
 
 file_rotation_option
   : KW_SIZE '(' nonnegative_integer ')' { file_rotation_set_size(last_file_rotation_plugin, $3); }
+  | KW_INTERVAL '(' string ')' { log_rotation_set_interval(last_file_rotation_plugin, $3); free($3); }
```
The PR still can not be built. `log_rotation_set_interval` -> `file_rotation_set_interval`

shikharvashistha replied:
The file where I made changes couldn't be pushed due to git issues.

MrAnno commented:
There is one remaining CMake build issue in the CI:

```
[ 80%] Building C object syslog-ng-ctl/CMakeFiles/syslog-ng-ctl.dir/syslog-ng-ctl.c.o
/__w/syslog-ng/syslog-ng/modules/file-rotation/file-rotation.c:25:10: fatal error: 'modules/affile/file-signals.h' file not found
#include "modules/affile/file-signals.h"
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

You can fix this by adding the following to `modules/file-rotation/CMakeLists.txt`:

```diff
--- a/modules/file-rotation/CMakeLists.txt
+++ b/modules/file-rotation/CMakeLists.txt
@@ -15,6 +15,7 @@ add_module(
   TARGET file_rotation
   GRAMMAR file-rotation-grammar
   SOURCES ${FILE_ROTATION_SOURCES}
+  INCLUDES ${PROJECT_SOURCE_DIR}
 )
```

shikharvashistha commented:
> There is one remaining CMake build issue in the CI:
> 
> ```
> [ 80%] Building C object syslog-ng-ctl/CMakeFiles/syslog-ng-ctl.dir/syslog-ng-ctl.c.o
> /__w/syslog-ng/syslog-ng/modules/file-rotation/file-rotation.c:25:10: fatal error: 'modules/affile/file-signals.h' file not found
> #include "modules/affile/file-signals.h"
>          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> ```
> 
> You can fix this by adding the following to `modules/file-rotation/CMakeLists.txt`:
> 
> ```diff
> --- a/modules/file-rotation/CMakeLists.txt
> +++ b/modules/file-rotation/CMakeLists.txt
> @@ -15,6 +15,7 @@ add_module(
>    TARGET file_rotation
>    GRAMMAR file-rotation-grammar
>    SOURCES ${FILE_ROTATION_SOURCES}
> +  INCLUDES ${PROJECT_SOURCE_DIR}
>  )
> ```

Yeah I've added it to CMakeLists.txt

shikharvashistha committed:
affile: Assigned AFFile's signal_slot_connetor to logproto's

This commit assigns the AFFileDestDriver signal_slot_connector to the logproto's file writer struct

Signed-off-by: Shikhar Vashistha <51105234+shikharvashistha@users.noreply.github.com>
